### DISCUSSION TITLE
Pure rust mysql client

### DISCUSSION THREAD

**Discussion by pickfire** [diesel-rs/diesel#2635] [Ideas]
Maybe it would be good to have a pure rust mysql client? sqlx already have it. https://docs.rs/mysql/20.1.0/mysql/index.html

**Comment by weiznich** [comment]
As this is a feature request: Could you add answers to the following questions from our contribution guidelines?


* What do you want to do, and how do you expect Diesel to support you with that?
* How might this be added to Diesel?
* What are possible alternatives?
* Are there any disadvantages?

Additionally for this request it may be important to elaborate the following points:

* Why this needs to be part of diesel itself and cannot be implemented as third party crate?
* Are there potential performance regressions and how would you handle them? ([Our benchmarks](https://github.com/diesel-rs/diesel/pull/2507#issuecomment-704286749) indicate that both, the mysql crate you've linked above and `sqlx` with the mysql backend are much less performant than diesel + libmysqlclient)

**Comment by pickfire** [comment]
> What do you want to do, and how do you expect Diesel to support you with that?

I don't need it, since I am just experimenting with diesel by trying out foot guns I faced in gorm. I am thinking this may be easier for others that wanted to easily get started with diesel since using pure rust does not require any libraries.

> How might this be added to Diesel?

Like a feature `pure-mysql` probably, I haven't looked into the details yet.

> What are possible alternatives?

sqlx is a possible alternative to diesel but both have very different design decisions, of course current mysql works fine and well, but cross-platform wise I think using pure rust backend will make things easier, not more performant though and may result in more complexity.

> Are there any disadvantages?

Yes, complexity wise, in increases more burden on maintainer to keep up to date with both libmysqlclient and rust mysql crate, and issues within rust mysql crate may propagate to diesel. Performance wise, if rust mysql is able to perform as well as mysql I don't think it is an issue but I haven't looked much into that crate, but that crate provides a pool connection, not sure how well it integrates with r2d2 and deadpool and such.

> Why this needs to be part of diesel itself and cannot be implemented as third party crate?

I didn't know this could be implemented as a third party crate, probably possible.

> Are there potential performance regressions and how would you handle them? (Our benchmarks indicate that both, the mysql crate you've linked above and sqlx with the mysql backend are much less performant than diesel + libmysqlclient)

I am not sure if the performance difference is caused by sqlx or mysql backend, it could be that it is caused by sqlx and not mysql backend. But as always, I am not suggesting to swap out libmysqlclient but rather provide another backend for mysql using pure rust. Since our mysql stuff is in place, it will probably not be as much work as adding a different database.

**Reply by weiznich** [reply]
> Yes, complexity wise, in increases more burden on maintainer to keep up to date with both libmysqlclient and rust mysql crate, and issues within rust mysql crate may propagate to diesel.

One of my current issues while maintaining diesel is that I do not have that much time, so adding more complexity is definitively something that I want to avoid at least till there is another maintainer. 

>>    Why this needs to be part of diesel itself and cannot be implemented as third party crate?
> I didn't know this could be implemented as a third party crate, probably possible.

If someone is interested in actually having a pure rust mysql connection doing that as third party crate would be the way to go, as for the maintenance reasons pointed out above. I'm happy to provide some guidance on how to implement such a crate, but maintenance must be done by someone else. If such a crate has reached a certain point of stability we could talk about up streaming the effort to diesel itself.
As for the how to implement such a crate: You would need to implement [`Connection`](http://docs.diesel.rs/master/diesel/prelude/trait.Connection.html) for your custom connection type. 


> I am not sure if the performance difference is caused by sqlx or mysql backend, it could be that it is caused by sqlx and not mysql backend.

sqlx, the mysql crate and diesel use all different code to talk to a mysql database. Both `sqlx` and the `mysql` crate implement their own pure rust driver (but different for each of them), while diesel uses libmysqlclient. The linked benchmark has numbers for all of them + as written above: The pure rust implementations are something like 2-5x times slower than libmysqlclient (at least according to my benchmark runs).
